Make skipping of LinuxCrossMingwTests more robust

petere opened the pull request:
The tests in LinuxCrossMingwTests look for a compiler 'x86_64-w64-mingw32-gcc' to decide whether to skip the tests.  This is done using shutil.which().  But the compiler binary could be a symlink to ccache (as is the case on macOS with Homebrew), in which case the compiler binary is found by shutil.which() but will not actually run successfully when called.  This will cause the test not to be skipped, but the test will then fail.

To fix, also try to run the compiler with --version to see if it will actually run successfully.  (This is similar to
AllPlatformTests.has_working_hg().)

dcbaker commented on unittests/linuxcrosstests.py:
```diff
@@ -116,7 +117,11 @@ def test_run_native_test(self):
+    # it will appear to exist but will fail when called.
+    return shutil.which('x86_64-w64-mingw32-gcc') and \
+        subprocess.call(['x86_64-w64-mingw32-gcc', '--version'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL) == 0 and \
+        not (is_windows() or is_cygwin())
```
Reading the docs I think `os.path.exists(shutil.which(…))` should be sufficient. It’s supposed to resolve symlinks correctly

eli-schwartz replied:
Not if it's a symlink to ccache though.

That being said I think just running it should be enough. No need to also use shutil.which?

petere replied:
> That being said I think just running it should be enough. No need to also use shutil.which?

done that way

eli-schwartz replied:
Pretty sure you do have to catch FileNotFoundError, which doesn't involve stat'ing the filesystem to see if it exists like shutil.which does, but is still some code to include.

eli-schwartz replied:
(This is probably a good thing since the logic was getting quite hairy and nested. :P)

petere replied:
When I try this in isolation I do not get an exception:

```python
import subprocess

ret = subprocess.call(['x86_64-w64-mingw32-gcc', '--version'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
print(ret)
```

just prints `1` without an exception.

petere replied:
Sorry, that didn't make any sense. I have pushed a new version with the exception check.

Personally, I don't think that turned out to be better than the version that checks `shutil.which` followed by `subprocess.call`, but I'm happy to oblige.

I also noticed that the `is_windows()` check in the `skipUnless` right below is redundant.

Finally, there is a similar issue in `should_run_cross_arm_tests` in the same file. It doesn't affect my system, but it would probably make sense to keep all these "is the compiler there" checks similar?

eli-schwartz replied:
> Personally, I don't think that turned out to be better than the version that checks `shutil.which` followed by `subprocess.call`, but I'm happy to oblige.

Hmm, that version had chained `and`s running across three lines, and a comment that was distanced from the line it was commenting on. Interesting that you prefer that.

petere replied:
> Finally, there is a similar issue in `should_run_cross_arm_tests` in the same file. It doesn't affect my system, but it would probably make sense to keep all these "is the compiler there" checks similar?

I have also changed that function in the updated version of the patch. This is complete from my perspective now.

petere committed:
Make skipping of LinuxCrossMingwTests more robust

The tests in LinuxCrossMingwTests look for a compiler
'x86_64-w64-mingw32-gcc' to decide whether to skip the tests.  This is
done using shutil.which().  But the compiler binary could be a symlink
to ccache (as is the case on macOS with Homebrew), in which case the
compiler binary is found by shutil.which() but will not actually run
successfully when called.  This will cause the tests not to be
skipped, but the tests will then fail.

To fix, instead try to run the compiler with --version to see if it
will actually run successfully.

Also adjust should_run_cross_arm_tests() in the same way, for
consistency.

petere commented:
Is anything else needed from me for this?
